File: mobile/lib/features/b2b/domain/entities/b2b_order.dart
import 'order_item.dart';
import 'order_status.dart';
import 'payment_status.dart';

class B2BOrder {
  final String id;
  final String reference;
  final String wholesalerId;
  final String wholesalerName;
  final String retailerId;
  final String retailerName;
  final OrderStatus status;
  final String statusName;
  final PaymentStatus paymentStatus;
  final String paymentStatusName;
  final List<OrderItem> items;
  final int itemCount;
  final double subtotal;
  final double discountAmount;
  final double totalAmount;
  final double amountPaid;
  final double amountDue;
  final bool useCredit;
  final String? notes;
  final String? deliveryAddress;
  final DateTime? expectedDeliveryDate;
  final DateTime createdAt;
  final DateTime? confirmedAt;
  final DateTime? shippedAt;
  final DateTime? deliveredAt;
  final DateTime? cancelledAt;
  final String? cancellationReason;

  const B2BOrder({
    required this.id,
    required this.reference,
    required this.wholesalerId,
    required this.wholesalerName,
    required this.retailerId,
    required this.retailerName,
    required this.status,
    required this.statusName,
    required this.paymentStatus,
    required this.paymentStatusName,
    required this.items,
    required this.itemCount,
    required this.subtotal,
    this.discountAmount = 0,
    required this.totalAmount,
    this.amountPaid = 0,
    required this.amountDue,
    this.useCredit = false,
    this.notes,
    this.deliveryAddress,
    this.expectedDeliveryDate,
    required this.createdAt,
    this.confirmedAt,
    this.shippedAt,
    this.deliveredAt,
    this.cancelledAt,
    this.cancellationReason,
  });

  B2BOrder copyWith({
    String? id,
    String? reference,
    String? wholesalerId,
    String? wholesalerName,
    String? retailerId,
    String? retailerName,
    OrderStatus? status,
    String? statusName,
    PaymentStatus? paymentStatus,
    String? paymentStatusName,
    List<OrderItem>? items,
    int? itemCount,
    double? subtotal,
    double? discountAmount,
    double? totalAmount,
    double? amountPaid,
    double? amountDue,
    bool? useCredit,
    String? notes,
    String? deliveryAddress,
    DateTime? expectedDeliveryDate,
    DateTime? createdAt,
    DateTime? confirmedAt,
    DateTime? shippedAt,
    DateTime? deliveredAt,
    DateTime? cancelledAt,
    String? cancellationReason,
  }) {
    return B2BOrder(
      id: id ?? this.id,
      reference: reference ?? this.reference,
      wholesalerId: wholesalerId ?? this.wholesalerId,
      wholesalerName: wholesalerName ?? this.wholesalerName,
      retailerId: retailerId ?? this.retailerId,
      retailerName: retailerName ?? this.retailerName,
      status: status ?? this.status,
      statusName: statusName ?? this.statusName,
      paymentStatus: paymentStatus ?? this.paymentStatus,
      paymentStatusName: paymentStatusName ?? this.paymentStatusName,
      items: items ?? this.items,
      itemCount: itemCount ?? this.itemCount,
      subtotal: subtotal ?? this.subtotal,
      discountAmount: discountAmount ?? this.discountAmount,
      totalAmount: totalAmount ?? this.totalAmount,
      amountPaid: amountPaid ?? this.amountPaid,
      amountDue: amountDue ?? this.amountDue,
      useCredit: useCredit ?? this.useCredit,
      notes: notes ?? this.notes,
      deliveryAddress: deliveryAddress ?? this.deliveryAddress,
      expectedDeliveryDate: expectedDeliveryDate ?? this.expectedDeliveryDate,
      createdAt: createdAt ?? this.createdAt,
      confirmedAt: confirmedAt ?? this.confirmedAt,
      shippedAt: shippedAt ?? this.shippedAt,
      deliveredAt: deliveredAt ?? this.deliveredAt,
      cancelledAt: cancelledAt ?? this.cancelledAt,
      cancellationReason: cancellationReason ?? this.cancellationReason,
    );
  }
}

File: backend/jufa-backend/src/main/java/ml/jufa/backend/b2b/entity/B2BOrder.java
package ml.jufa.backend.b2b.entity;

import jakarta.persistence.*;
import lombok.*;
import ml.jufa.backend.common.entity.BaseEntity;
import ml.jufa.backend.merchant.entity.MerchantProfile;
import ml.jufa.backend.merchant.entity.WholesalerRetailer;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "b2b_orders", indexes = {
    @Index(name = "idx_order_wholesaler", columnList = "wholesaler_id"),
    @Index(name = "idx_order_retailer", columnList = "retailer_id"),
    @Index(name = "idx_order_reference", columnList = "reference"),
    @Index(name = "idx_order_status", columnList = "status")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class B2BOrder extends BaseEntity {

    @Column(nullable = false, unique = true, length = 50)
    private String reference;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "wholesaler_id", nullable = false)
    private MerchantProfile wholesaler;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "retailer_id", nullable = false)
    private MerchantProfile retailer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "relation_id")
    private WholesalerRetailer relation;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private OrderStatus status = OrderStatus.PENDING;

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_status", nullable = false)
    @Builder.Default
    private PaymentStatus paymentStatus = PaymentStatus.PENDING;

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<OrderItem> items = new ArrayList<>();

    @Column(name = "subtotal", precision = 18, scale = 2)
    @Builder.Default
    private BigDecimal subtotal = BigDecimal.ZERO;

    @Column(name = "discount_amount", precision = 18, scale = 2)
    @Builder.Default
    private BigDecimal discountAmount = BigDecimal.ZERO;

    @Column(name = "total_amount", nullable = false, precision = 18, scale = 2)
    @Builder.Default
    private BigDecimal totalAmount = BigDecimal.ZERO;

    @Column(name = "amount_paid", precision = 18, scale = 2)
    @Builder.Default
    private BigDecimal amountPaid = BigDecimal.ZERO;

    @Column(name = "use_credit")
    @Builder.Default
    private Boolean useCredit = false;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Column(name = "delivery_address", columnDefinition = "TEXT")
    private String deliveryAddress;

    @Column(name = "expected_delivery_date")
    private LocalDate expectedDeliveryDate;

    @Column(name = "confirmed_at")
    private LocalDateTime confirmedAt;

    @Column(name = "shipped_at")
    private LocalDateTime shippedAt;

    @Column(name = "delivered_at")
    private LocalDateTime deliveredAt;

    @Column(name = "cancelled_at")
    private LocalDateTime cancelledAt;

    @Column(name = "cancellation_reason")
    private String cancellationReason;

    public void addItem(OrderItem item) {
        items.add(item);
        item.setOrder(this);
        recalculateTotals();
    }

    public void removeItem(OrderItem item) {
        items.remove(item);
        item.setOrder(null);
        recalculateTotals();
    }

    public void recalculateTotals() {
        this.subtotal = items.stream()
                .map(OrderItem::getLineTotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
        this.totalAmount = subtotal.subtract(discountAmount != null ? discountAmount : BigDecimal.ZERO);
    }

    public BigDecimal getAmountDue() {
        return totalAmount.subtract(amountPaid != null ? amountPaid : BigDecimal.ZERO);
    }

    public void confirm() {
        this.status = OrderStatus.CONFIRMED;
        this.confirmedAt = LocalDateTime.now();
    }

    public void ship() {
        this.status = OrderStatus.SHIPPED;
        this.shippedAt = LocalDateTime.now();
    }

    public void deliver() {
        this.status = OrderStatus.DELIVERED;
        this.deliveredAt = LocalDateTime.now();
    }

    public void cancel(String reason) {
        this.status = OrderStatus.CANCELLED;
        this.cancelledAt = LocalDateTime.now();
        this.cancellationReason = reason;
    }
}
